feat: Pod creation settings

vv9k opened the pull request:
This PR adds some settings to the pod creation page as well as adding the ability to specify display labels of `PdsKeyValRow`

![image](https://user-images.githubusercontent.com/46892771/193661112-4cc6e07f-963f-449a-b678-7181399efc34.png)

![image](https://user-images.githubusercontent.com/46892771/193661161-cce6186d-badf-41f6-aa57-146e717286ee.png)

marhkb commented on src/view/pod/creation_page.rs:
```diff
@@ -45,8 +53,24 @@ mod imp {
+        pub(super) devices_list_box: TemplateChild<gtk::ListBox>,
+        #[template_child]
+        pub(super) disable_hosts_switch: TemplateChild<gtk::Switch>,
+        #[template_child]
+        pub(super) disable_hosts_row: TemplateChild<adw::ActionRow>,
```
```suggestion
```

marhkb commented on data/resources/ui/pod/creation-page.ui:
```diff
@@ -114,6 +163,49 @@
+                                    <child>
+                                      <object class="AdwPreferencesGroup">
+                                        <property name="title" translatable="yes">Hosts</property>
+
+                                        <child>
+                                          <object class="AdwActionRow" id="disable_hosts_row">
+                                            <property name="activatable_widget">disable_hosts_switch</property>
+                                            <property name="title" translatable="yes">Disable Hosts management</property>
+                                            <child>
+                                              <object class="GtkSwitch" id="disable_hosts_switch">
+                                                <property name="valign">center</property>
+                                                <property name="action-name">pod.toggle-hosts</property>
+                                              </object>
+                                            </child>
+                                          </object>
+                                        </child>
```
It looks better if we use a switch in the header suffix. I would also change the behavior in a way that the hosts become enabled if the switch is active instead of disable it.

![grafik](https://user-images.githubusercontent.com/3630213/193688552-41c38116-5f63-4509-ac49-51a2c70957e5.png)

```suggestion
                                        <property name="header-suffix">
                                          <object class="GtkBox">
                                            <property name="spacing">6</property>

                                            <child>
                                              <object class="GtkLabel">
                                                <property name="label" translatable="yes">Enabled</property>
                                              </object>
                                            </child>

                                            <child>
                                              <object class="GtkSwitch" id="enable_hosts_switch">
                                                <property name="valign">center</property>
                                                <property name="action-name">pod.toggle-hosts</property>
                                              </object>
                                            </child>

                                          </object>
                                        </property>
```

vv9k replied:
That is actually what I was aiming for initially but I'm just very unfamiliar with GTK :)

marhkb commented:
@vv9k Thank you very much. I've had a quick look at the code and left a small review. This already looks quite good.
Will have a closer look at it within the next days.

vv9k commented:
Fixed outstanding issues and rebased on main.

One other thing that I couldn't figure out was how to make the `container path` and `host path` text inputs the same width in the `DeviceRow`. Seems that the width of the `host path` text input is wider because of the first row that contains the switches.

marhkb commented on src/view/key_val/row.rs:
```diff
@@ -67,16 +89,30 @@ mod imp {
+
+        fn constructed(&self, obj: &Self::Type) {
+            self.parent_constructed(obj);
+            obj.imp()
+                .key_entry
+                .set_property("placeholder-text", self.key_label.borrow().as_str());
+            obj.imp()
+                .value_entry
+                .set_property("placeholder-text", self.value_label.borrow().as_str());
```
As we delegate to `Entry::placeholder_text` to write the property, we watch for changes here to notify about our property.

```suggestion
            self.key_entry
                .connect_placeholder_text_notify(clone!(@weak obj => move |_| {
                    obj.notify("key-label");
                }));
            self.key_entry
                .connect_placeholder_text_notify(clone!(@weak obj => move |_| {
                    obj.notify("value-label");
                }));
```

marhkb commented on data/resources/ui/device/row.ui:
```diff
@@ -0,0 +1,108 @@
+<?xml version="1.0" encoding="UTF-8"?>
+<interface>
+
+  <object class="GtkSizeGroup">
+    <property name="mode">horizontal</property>
+    <widgets>
+      <widget name="mode_box"/>
+      <widget name="host_path_entry"/>
+    </widgets>
+  </object>
+
+  <object class="GtkSizeGroup">
+    <property name="mode">horizontal</property>
+    <widgets>
+      <widget name="container_path_entry"/>
+    </widgets>
+  </object>
```
> One other thing that I couldn't figure out was how to make the container path and host path text inputs the same width in the DeviceRow. Seems that the width of the host path text input is wider because of the first row that contains the switches.

You're correct. The mode box makes the host path entry too wide. Removing all groups will fix it.

```suggestion
```


With the other suggestions below, the end result will look like

![grafik](https://user-images.githubusercontent.com/3630213/193918606-ae9c8caa-5a32-444e-b13d-cae63b325754.png)

marhkb commented on data/resources/ui/device/row.ui:
```diff
@@ -0,0 +1,108 @@
+        <child>
+          <object class="GtkBox" id="mode_box">
+            <property name="hexpand">True</property>
+            <property name="margin-start">9</property>
+            <property name="spacing">9</property>
```
Wider margins make this look better.

```suggestion
        <property name="margin-bottom">18</property>
        <property name="margin-end">12</property>
        <property name="margin-start">12</property>
        <property name="margin-top">18</property>
        <property name="spacing">12</property>
```

vv9k replied:
Looks much better!

marhkb commented on data/resources/ui/device/row.ui:
```diff
@@ -0,0 +1,108 @@
+            <property name="margin-start">9</property>
+            <property name="spacing">9</property>
+
+            <child>
+              <object class="GtkLabel">
+                <property name="label" translatable="yes">Readable</property>
+              </object>
+            </child>
+
+            <child>
+              <object class="GtkSwitch" id="readable_switch">
+                <property name="valign">center</property>
+              </object>
+            </child>
+
+            <child>
+              <object class="GtkLabel">
+                <property name="label" translatable="yes">Writable</property>
+              </object>
+            </child>
+
+            <child>
+              <object class="GtkSwitch" id="writable_switch">
+                <property name="valign">center</property>
+              </object>
+            </child>
+
+
+            <child>
+              <object class="GtkLabel">
+                <property name="label" translatable="yes">mknod</property>
+              </object>
+            </child>
+
+            <child>
+              <object class="GtkSwitch" id="mknod_switch">
+                <property name="valign">center</property>
+              </object>
+            </child>
```
The switches are better separated if we put them in single boxes together with their labels. Also `id="mode_box"` can be removed.

```suggestion
        <child>
          <object class="GtkBox">
            <property name="margin-start">6</property>
            <property name="spacing">18</property>

            <child>
              <object class="GtkBox">
                <property name="spacing">6</property>

                <child>
                  <object class="GtkLabel">
                    <property name="label" translatable="yes">Readable</property>
                  </object>
                </child>

                <child>
                  <object class="GtkSwitch" id="readable_switch">
                    <property name="valign">center</property>
                  </object>
                </child>

              </object>
            </child>

            <child>
              <object class="GtkBox">
                <property name="spacing">6</property>

                <child>
                  <object class="GtkLabel">
                    <property name="label" translatable="yes">Writable</property>
                  </object>
                </child>

                <child>
                  <object class="GtkSwitch" id="writable_switch">
                    <property name="valign">center</property>
                  </object>
                </child>

              </object>
            </child>

            <child>
              <object class="GtkBox">
                <property name="spacing">6</property>

                <child>
                  <object class="GtkLabel">
                    <property name="label" translatable="yes">mknod</property>
                  </object>
                </child>

                <child>
                  <object class="GtkSwitch" id="mknod_switch">
                    <property name="valign">center</property>
                  </object>
                </child>

              </object>
            </child>

          </object>
        </child>
```

marhkb commented on src/view/pod/creation_page.rs:
```diff
@@ -21,8 +21,13 @@ use crate::view;
 
 const ACTION_CREATE: &str = "pod-creation-page.create";
 const ACTION_ADD_LABEL: &str = "pod.add-label";
+const ACTION_ADD_HOST: &str = "pod.add-host";
```
I started to use the widget name as the first part of the action name (forgot to rename those here...).
This makes it easier to understand in which widget the action is defined (as actions are searched in the root direction of the widget tree and there are some places where widgets use actions of parent widgets).

The same applies to other actions that were added in this PR.

At some point, I will rename the remaining actions in a separate PR.

```suggestion
const ACTION_ADD_HOST: &str = "pod-creation-page.add-host";
```

marhkb commented on data/resources/ui/pod/creation-page.ui:
```diff
@@ -114,6 +163,55 @@
+
+                                            <child>
+                                              <object class="GtkSwitch" id="enable_hosts_switch">
+                                                <property name="valign">center</property>
```
Else, the switch is not active while the "+" Button is visible.

```suggestion
                                                <property name="active">True</property>
                                                <property name="valign">center</property>
```

marhkb commented on data/resources/ui/pod/creation-page.ui:
```diff
@@ -114,6 +163,56 @@
+                                              <object class="GtkSwitch" id="enable_hosts_switch">
+                                                <property name="active">true</property>
+                                                <property name="valign">center</property>
+                                                <property name="action-name">pod.toggle-hosts</property>
```
```suggestion
                                                <property name="action-name">pod-creation-page.toggle-hosts</property>
```

vv9k commented on data/resources/ui/pod/creation-page.ui:
```diff
@@ -89,12 +89,61 @@
+                                            <child>
+                                              <object class="GtkSwitch" id="disable_resolv_switch">
+                                                <property name="valign">center</property>
+                                                <property name="action-name">pod.toggle-resolv</property>
```
Completely forgot about that, updated

marhkb commented on src/view/key_val/row.rs:
```diff
@@ -91,11 +128,23 @@ glib::wrapper! {
 impl From<&model::KeyVal> for Row {
     fn from(key_val: &model::KeyVal) -> Self {
-        glib::Object::new(&[("key-val", &key_val)]).expect("Failed to create PdsKeyValRow")
+        Row::new("Key", "Value", key_val)
```
```suggestion
        Row::new(gettext("Key"), gettext("Value"), key_val)
```

marhkb commented on src/view/pod/creation_page.rs:
```diff
@@ -177,6 +214,67 @@ mod imp {
+            self.hosts_list_box
+                .bind_model(Some(&*self.hosts.borrow()), |item| {
+                    view::KeyValRow::new(
+                        "Hostname",
+                        "IP",
```
```suggestion
                        gettext("Hostname"),
                        gettext("IP"),
```

marhkb commented:
LGTM. Stumbled across 2 last remaining things regarding i18n.

vv9k commented:
Fixed remaining issues

marhkb commented:
Wonderful. Thanks very much again.
